src/dataqueue/queue.cc

```diff
@@ -0,0 +1,1092 @@
+// check is also async, so there's a natural race condition there where the
+// file could be modified between the stat and actual read calls. That's
+// a tolerable risk here. While FdEntry is considered idempotent, this race
+// means that it is indeed possible for multiple reads to return different
+// results if the file just happens to get modified.
+class FdEntry final : public EntryImpl {
+  // TODO(@jasnell, @flakey5):
+  // * This should only allow reading from regular files. No directories, no
+  // pipes, etc.
+  // * The reader should support accepting the buffer(s) from the pull, if any.
+  // It should
+  //   only allocate a managed buffer if the pull doesn't provide any.
+  // * We might want to consider making the stat on each read sync to eliminate
+  // the race
+  //   condition described in the comment above.
+ public:
+  static std::unique_ptr<FdEntry> Create(Environment* env, Local<Value> path) {
+    // We're only going to create the FdEntry if the file exists.
+    uv_fs_t req;
+    auto cleanup = OnScopeLeave([&] { uv_fs_req_cleanup(&req); });
+
+    auto buf = std::make_shared<BufferValue>(env->isolate(), path);
+    if (uv_fs_stat(nullptr, &req, buf->out(), nullptr) < 0) return nullptr;
+
+    return std::make_unique<FdEntry>(
+        env, std::move(buf), req.statbuf, 0, req.statbuf.st_size);
+  }
+
+  FdEntry(Environment* env,
+          std::shared_ptr<BufferValue> path_,
+          uv_stat_t stat,
+          uint64_t start,
+          uint64_t end)
+      : env_(env),
+        path_(std::move(path_)),
+        stat_(stat),
+        start_(start),
+        end_(end) {}
+
+  std::shared_ptr<DataQueue::Reader> get_reader() override {
+    return ReaderImpl::Create(this);
+  }
+
+  std::unique_ptr<Entry> slice(
+      uint64_t start, std::optional<uint64_t> end = std::nullopt) override {
+    uint64_t new_start = start_ + start;
+    uint64_t new_end = end_;
+    if (end.has_value()) {
+      new_end = std::min(end.value() + start, new_end);
+    }
+
+    CHECK(new_start >= start_);
+    CHECK(new_end <= end_);
+
+    return std::make_unique<FdEntry>(env_, path_, stat_, new_start, new_end);
+  }
+
+  std::optional<uint64_t> size() const override { return end_ - start_; }
+
+  bool is_idempotent() const override { return true; }
+
+  Environment* env() const { return env_; }
+
+  SET_NO_MEMORY_INFO()
+  SET_MEMORY_INFO_NAME(FdEntry)
+  SET_SELF_SIZE(FdEntry)
+
+ private:
+  Environment* env_;
+  std::shared_ptr<BufferValue> path_;
+  uv_stat_t stat_;
+  uint64_t start_ = 0;
+  uint64_t end_ = 0;
+
+  bool is_modified(const uv_stat_t& other) {
+    return other.st_size != stat_.st_size ||
+           other.st_mtim.tv_nsec != stat_.st_mtim.tv_nsec;
+  }
+
+  static bool CheckModified(FdEntry* entry, int fd) {
+    uv_fs_t req;
+    auto cleanup = OnScopeLeave([&] { uv_fs_req_cleanup(&req); });
+    if (uv_fs_fstat(nullptr, &req, fd, nullptr) < 0) return true;
```

Comment:
Similarly here, we should probably not be doing synchronous fs I/O under the hood if we can avoid it

Reply:
As above, for now, I'd prefer to keep these as is and revisit later.